### DISCUSSION TITLE
Gaming and 3D graphics

### DISCUSSION THREAD

**Discussion by freever** [dotnet/maui#196] [General]
Hi folks. Where does gaming and 3D graphics fit in to the MAUI roadmap, if at all? 

Referring to [this](https://docs.microsoft.com/en-us/xamarin/graphics-games/game-development/) fairly old overview, the options are MonoGame and UrhoSharp. I have been using UrhoSharp. The engine itself is really neat and the integration with Xamarin.Forms is fantastic. But the UrhoSharp project seems dormant, and I am really worried about being left on a stale and unsupported platform. I don't want to move to a game engine proper because I like the ability to embed 3D graphics within a traditional app structure as it gives me the best of both worlds. 

Will UrhoSharp be coming along for the MAUI ride, or is it basically dead? Is there an alternative approach I should be taking for 3D graphics and gaming in the MAUI world? 

I'd really appreciate some guidance.

**Comment by Happypig375** [comment]
MonoGame is still alive and ongoing.

**Reply by freever** [reply]
@Happypig375 It does indeed seem to be alive and kicking. However MonoGame does not offer a way to embed a game screen inside a XF/MAUI app. Like Unity3D it's an either-or proposition: either build a "game" or build an "app". 

What I am looking for is a way to embed a 3D graphics screen inside an app without having to use the gaming system to render all my UI, just as I can embed a 2D canvas using SkiaSharp. UrhoSharp is the only option I've found, and it works great. It's such a shame that it seems to have been left in the basement with a bag of crisps to die slowly in the dark with all of Microsoft's other red-headed step children. 

If MAUI is going to be the one-stop shop for developing apps of all kinds using .NET, I can't believe there isn't a solution for 3D graphics that doesn't require you to rewrite your whole app as a game, but that appears to be the case.

**Comment by charlesroddie** [comment]
Duplicate of https://github.com/dotnet/maui/issues/129

**Comment by jameskolts** [comment]
Wondering this as well.  Any update?

**Comment by asteroid7687** [comment]
Same question here. Any news?

**Reply by freever** [reply]
Lol at the moment I am doing my own 3D camera to calculate 2d screen coordinates from 3D vectors and rendering them with Skiasharp. Luckily I just need to do simple wireframes so it actually kinda works. Would love to see something like veldrid come out with a Maui control though

**Reply by pldeschamps** [reply]
That's a work around but your 3D camera won't use the GPU.
I face the same issue : I use urhosharp and I wonder what will happen when the code migrates from xamarin.forms to MAUI.

**Reply by KieranDevvs** [reply]
https://github.com/dotnet/Silk.NET
It uses GLFW underneath, to handle your cross platform windowing needs for 3D GPU accelerated graphics.

**Comment by LazyCoderSuhe** [comment]
同样的问题等待 微软给出答案 3D 在MAUI 中选择那个?

**Comment by ksy-3528** [comment]
I am also using UrhoSharp and wondering the MAUI team will support that on MAUI. I wonder how MAUI team would migrate so many libraries from Xamarin to MAUI. I hope that they migrate at least those libraries supported by Microsoft.(You know that Urhosharp was supported by Microsoft). I also hope that MAUI support the native binding ASAP ( aar, framework, dll binding).

**Comment by Xyncgas** [comment]
If I am making games, I will have the money to talk to steams and Nvidia, and I can play any games on the cloud via Nvidia GeforceNow and you can even build your own front end before launching the game

Because cloud gaming can run on browser and there's a webview control for MAUI, similarly you can run any games that runs on the browser using webassembly or flash or javascript.

Alternatively, you can just build the front end using MAUI, then launch your own game locally via process invoke

the possibility is endless, however I doubt you can make games with MAUI or .NET in general, but fine if you want to make games using c# there is always unity, you can make games using unity just as usually or any other engines of your choice I think developers would much prefer drawing a box using gaming engine than XAML, then run the client since you have local access

I can't wait to see the games you people make

**Comment by xtuzy** [comment]
i migrate veldrid example to maui, see https://github.com/xtuzy/veldrid.
